File: backend-go/internal/registry/nuget.go
package registry

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

var (
	nugetIDRe      = regexp.MustCompile(`^[A-Za-z0-9]([A-Za-z0-9._-]*[A-Za-z0-9])?$`)
	nugetVersionRe = regexp.MustCompile(`^\d+\.\d+(\.\d+)?(\.\d+)?(-[0-9A-Za-z.\-]+)?(\+[0-9A-Za-z.\-]+)?$`)
)

// NuGet — плагин менеджера nuget. Формат записи: Id@version.
type NuGet struct {
	BaseURL string
	HTTP    Doer
}

func (*NuGet) Code() string         { return "nuget" }
func (*NuGet) Title() string        { return "NuGet (.NET)" }
func (*NuGet) EntryFormat() string  { return "Id@version" }
func (*NuGet) OSVEcosystem() string { return "NuGet" }

func (*NuGet) NormalizeName(name string) string {
	return strings.ToLower(strings.TrimSpace(name))
}
func (*NuGet) NormalizeVersion(version string) string {
	return strings.ToLower(strings.TrimSpace(version))
}
func (*NuGet) DisplayName(name string) string { return strings.TrimSpace(name) }

func (p *NuGet) SplitEntry(entry string) (string, string, error) {
	text := strings.TrimSpace(entry)
	idx := strings.LastIndex(text, "@")
	if idx < 0 {
		return "", "", invalidFormat(p.EntryFormat(),
			"«%s» не соответствует формату nuget. Ожидается: Id@version "+
				"(например, Newtonsoft.Json@13.0.3)", text)
	}
	return strings.TrimSpace(text[:idx]), strings.TrimSpace(text[idx+1:]), nil
}

func (p *NuGet) ValidateName(name string) error {
	if len(name) > 512 {
		return invalidFormat(p.EntryFormat(), "Имя пакета длиннее 512 символов")
	}
	if !nugetIDRe.MatchString(name) {
		return invalidFormat(p.EntryFormat(),
			"Недопустимый Id пакета nuget: «%s» (буквы, цифры, «.», «-», «_»)", name)
	}
	return nil
}

func (p *NuGet) ValidateVersion(version string) error {
	if len(version) > 128 {
		return invalidFormat(p.EntryFormat(), "Версия длиннее 128 символов")
	}
	if !nugetVersionRe.MatchString(strings.TrimSpace(version)) {
		return invalidFormat(p.EntryFormat(),
			"Версия «%s» не соответствует формату nuget (например, 13.0.3, 6.0.0-preview.5)", version)
	}
	return nil
}

// nugetRegistration — ответ registration-эндпоинта. catalogEntry бывает и
// объектом, и ссылкой на него — отсюда json.RawMessage.
type nugetRegistration struct {
	CatalogEntry   json.RawMessage `json:"catalogEntry"`
	PackageContent string          `json:"packageContent"`
}

type nugetCatalogEntry struct {
	Published         string `json:"published"`
	LicenseExpression string `json:"licenseExpression"`
	LicenseURL        string `json:"licenseUrl"`
	PackageContent    string `json:"packageContent"`
	Listed            *bool  `json:"listed"`
}

func (p *NuGet) FetchMetadata(ctx context.Context, ref Ref) (Metadata, error) {
	url := fmt.Sprintf("%s/v3/registration5-semver1/%s/%s.json", p.BaseURL, ref.Name, ref.Version)
	var payload nugetRegistration
	if err := getJSON(ctx, p.HTTP, url, "application/json", &payload); err != nil {
		if err == ErrNotFound {
			return Metadata{}, fmt.Errorf("%w: пакет %s@%s отсутствует в реестре nuget",
				ErrNotFound, ref.DisplayName, ref.RawVersion)
		}
		return Metadata{}, err
	}

	entry, err := p.resolveCatalogEntry(ctx, payload.CatalogEntry)
	if err != nil {
		return Metadata{}, err
	}

	content := payload.PackageContent
	if content == "" {
		content = entry.PackageContent
	}
	if content == "" {
		content = fmt.Sprintf("%s/v3-flatcontainer/%s/%s/%s.%s.nupkg",
			p.BaseURL, ref.Name, ref.Version, ref.Name, ref.Version)
	}

	meta := Metadata{
		Name:             ref.Name,
		Version:          ref.Version,
		PublishedAt:      parseTime(entry.Published),
		ArtifactURL:      content,
		ArtifactFilename: fmt.Sprintf("%s.%s.nupkg", ref.Name, ref.Version),
		LicenseSPDX:      NormalizeSPDX(entry.LicenseExpression),
		// NuGet отдаёт хеш только внутри манифеста .nupkg — sha256 считаем
		// сами по скачанному артефакту.
		Yanked: entry.Listed != nil && !*entry.Listed,
	}
	meta.LicenseRaw = entry.LicenseExpression
	if meta.LicenseRaw == "" {
		meta.LicenseRaw = entry.LicenseURL
	}
	return meta, nil
}

// resolveCatalogEntry разбирает catalogEntry: объектом — разбираем на месте,
// строкой — это ссылка, ходим за ней. Недоступность каталога по ссылке не
// валит шаг: путь к артефакту и так известен, а лицензию определит юрист.
func (p *NuGet) resolveCatalogEntry(ctx context.Context, raw json.RawMessage) (nugetCatalogEntry, error) {
	var entry nugetCatalogEntry
	if len(raw) == 0 {
		return entry, nil
	}
	var link string
	if err := json.Unmarshal(raw, &link); err == nil {
		if err := getJSON(ctx, p.HTTP, link, "application/json", &entry); err != nil {
			return nugetCatalogEntry{}, nil //nolint:nilerr // см. комментарий выше
		}
		return entry, nil
	}
	if err := json.Unmarshal(raw, &entry); err != nil {
		return entry, fmt.Errorf("catalogEntry реестра nuget не разобран: %w", err)
	}
	return entry, nil
}

func (p *NuGet) InstallCommand(ref Ref, baseURL, repo string) string {
	source := fmt.Sprintf("%s/repository/%s/index.json", strings.TrimRight(baseURL, "/"), repo)
	return fmt.Sprintf("dotnet nuget add source %s -n internal && dotnet add package %s -v %s -s %s",
		source, ref.DisplayName, ref.RawVersion, source)
}

func (p *NuGet) ArtifactPath(ref Ref, filename string) string {
	return fmt.Sprintf("%s/%s/%s", ref.Name, ref.Version, filename)
}

File: backend/app/managers/nuget.py
"""Плагин менеджера nuget. Формат записи: `Id@version`."""

from __future__ import annotations

import re
from typing import Any

from app.core.config import get_settings
from app.core.errors import InvalidPackageFormat, UpstreamError
from app.core.http import client, request_with_retries
from app.managers.base import PackageManagerPlugin, PackageRef, ParsedEntry, RegistryMetadata
from app.managers.parsers.dotnet_files import (
    parse_csproj,
    parse_packages_config,
    parse_packages_lock_json,
)
from app.managers.utils import parse_iso8601, raw_deps_to_entries
from app.managers.versioning import compare_nuget
from app.services.spdx import normalize_spdx

_ID_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9._\-]*[A-Za-z0-9])?$")
_VERSION_RE = re.compile(r"^\d+(?:\.\d+){0,3}(?:-[0-9A-Za-z.\-]+)?(?:\+[0-9A-Za-z.\-]+)?$")


class NugetPlugin(PackageManagerPlugin):
    code = "nuget"
    title = "NuGet (.NET)"
    entry_format = "Id@version"
    dependency_files = ("packages.lock.json", "packages.config", "*.csproj")

    def normalize_name(self, name: str) -> str:
        return name.strip().lower()

    def normalize_version(self, version: str) -> str:
        return version.strip().lower()

    def split_entry(self, entry: str) -> tuple[str, str]:
        text = entry.strip()
        if "@" not in text:
            raise InvalidPackageFormat(
                f"«{text}» не соответствует формату nuget. Ожидается: Id@version "
                "(например, Newtonsoft.Json@13.0.3)",
                expected_format=self.entry_format,
            )
        name, _, version = text.rpartition("@")
        return name.strip(), version.strip()

    def validate_name(self, name: str) -> None:
        super().validate_name(name)
        if not _ID_RE.match(name):
            raise InvalidPackageFormat(
                f"Недопустимый Id пакета nuget: «{name}» (буквы, цифры, «.», «-», «_»)"
            )

    def validate_version(self, version: str) -> None:
        super().validate_version(version)
        if not _VERSION_RE.match(version):
            raise InvalidPackageFormat(
                f"Версия «{version}» не соответствует формату nuget (например, 13.0.3, 6.0.0-preview.5)"
            )

    def compare_versions(self, a: str, b: str) -> int:
        return compare_nuget(a, b)

    # ------------------------------------------------------------------ файлы
    def parse_dependency_file(self, filename: str, content: bytes) -> list[ParsedEntry]:
        base = filename.rsplit("/", 1)[-1]
        if base == "packages.lock.json":
            raw = parse_packages_lock_json(content)
        elif base == "packages.config":
            raw = parse_packages_config(content)
        elif base.endswith(".csproj"):
            raw = parse_csproj(content)
        else:
            raise InvalidPackageFormat(
                f"Файл «{base}» не поддерживается для nuget. Поддерживаются: "
                "packages.lock.json, packages.config, *.csproj"
            )
        return raw_deps_to_entries(self, raw)

    # ------------------------------------------------------------------ реестр
    def fetch_metadata(self, ref: PackageRef) -> RegistryMetadata:
        base = get_settings().registry_nuget_url.rstrip("/")
        pid, version = ref.name, ref.version
        url = f"{base}/v3/registration5-semver1/{pid}/{version}.json"
        with client() as http:
            resp = request_with_retries("nuget", lambda: http.get(url))
            if resp.status_code == 404:
                raise UpstreamError(
                    f"Пакет {ref.display_name}@{ref.raw_version} не найден в реестре nuget",
                    registry="nuget",
                )
            if resp.status_code >= 400:
                raise UpstreamError(f"Реестр nuget ответил {resp.status_code}", registry="nuget")
            payload: dict[str, Any] = resp.json()
        return self._parse_metadata(ref, payload, base)

    def _parse_metadata(self, ref: PackageRef, payload: dict[str, Any], base: str) -> RegistryMetadata:
        entry = payload.get("catalogEntry")
        if isinstance(entry, str):  # ссылка на каталог вместо inline-объекта
            with client() as http:
                resp = request_with_retries("nuget", lambda: http.get(entry))
            entry = resp.json() if resp.status_code < 400 else {}
        entry = entry or payload

        pid, version = ref.name, ref.version
        content_url = payload.get("packageContent") or entry.get("packageContent")
        if not content_url:
            content_url = f"{base}/v3-flatcontainer/{pid}/{version}/{pid}.{version}.nupkg"
        license_expr = entry.get("licenseExpression") or None
        return RegistryMetadata(
            name=ref.name,
            version=ref.version,
            published_at=parse_iso8601(entry.get("published")),
            license_spdx=normalize_spdx(license_expr),
            license_raw=license_expr or entry.get("licenseUrl"),
            artifact_url=content_url,
            artifact_filename=f"{pid}.{version}.nupkg",
            # NuGet отдаёт хеш только внутри .nupkg-манифеста; sha512 берётся из v3-flatcontainer
            checksum=None,
            checksum_algo=None,
            yanked=bool(entry.get("listed") is False),
            raw={"licenseUrl": entry.get("licenseUrl")} if entry.get("licenseUrl") else {},
        )

    # ------------------------------------------------------------------ публикация
    def publish_target(self, ref: PackageRef, filename: str) -> dict[str, str]:
        return {"format": "nuget", "filename": filename, "asset_field": "nuget.asset"}

    def install_command(self, ref: PackageRef, base_url: str, repo: str) -> str:
        source = f"{base_url.rstrip('/')}/repository/{repo}/index.json"
        return (
            f"dotnet nuget add source {source} -n internal && "
            f"dotnet add package {ref.display_name} -v {ref.raw_version} -s {source}"
        )
